subprojects/file-collections/src/main/java/org/gradle/api/internal/file/collections/GeneratedSingletonFileTree.java

```diff
@@ -42,16 +43,18 @@
     private final FileSystem fileSystem = FileSystems.getDefault();
 
     private final String fileName;
+    private final Consumer<String> beforeFileChange;
     private final Action<OutputStream> contentWriter;
 
-    public GeneratedSingletonFileTree(Factory<File> tmpDirSource, String fileName, Action<OutputStream> contentWriter) {
-        this(tmpDirSource, fileName, new PatternSet(), contentWriter);
+    public GeneratedSingletonFileTree(Factory<File> tmpDirSource, String fileName, Consumer<String> beforeFileChange, Action<OutputStream> contentWriter) {
```

Comment:
Let's add a proper listener type for this.

Reply:
Done as part of https://github.com/gradle/gradle/pull/11819.